docs/codeql/codeql-language-guides/analyzing-data-flow-in-ruby.rst

```diff
@@ -0,0 +1,389 @@
+    -  ``Concepts::Path::PathNormalization`` - a data-flow node that performs path normalization. This is often needed in order to safely access paths.
+    -  ``Concepts::CodeExecution`` - a data-flow node that dynamically executes Python code.
+    -  ``Concepts::SqlExecution`` - a data-flow node that executes SQL statements.
+    -  ``Concepts::HTTP::Server::RouteSetup`` - a data-flow node that sets up a route on a server.
+    -  ``Concepts::HTTP::Server::HttpResponse`` - a data-flow node that creates an HTTP response on a server.
+
+-  ``TaintTracking::Configuration`` - base class for custom global taint tracking analysis.
+
+Examples
+~~~~~~~~
+
+This query shows a data flow configuration that uses all network input as data sources:
+
+.. code-block:: ql
+
+    import codeql.ruby.DataFlow
+    import codeql.ruby.TaintTracking
+    import codeql.ruby.Concepts
+    import codeql.ruby.dataflow.RemoteFlowSources
+    
+    class RemoteToFileConfiguration extends TaintTracking::Configuration {
+      RemoteToFileConfiguration() { this = "RemoteToFileConfiguration" }
+    
+      override predicate isSource(DataFlow::Node source) { source instanceof RemoteFlowSource }
+    
+      override predicate isSink(DataFlow::Node sink) {
+        sink = any(FileSystemAccess fa).getAPathArgument()
+      }
+    }
+    
+    from DataFlow::Node input, DataFlow::Node fileAccess, RemoteToFileConfiguration config
+    where config.hasFlow(input, fileAccess)
+    select fileAccess, "This file access uses data from $@.", input, "user-controllable input."
+
+This data flow configuration tracks data flow from environment variables to opening files:
+
+.. code-block:: ql
+
+    import codeql.ruby.DataFlow
+    import codeql.ruby.controlflow.CfgNodes
+    import codeql.ruby.ApiGraphs
+    
+    class EnvironmentToFileConfiguration extends DataFlow::Configuration {
+      EnvironmentToFileConfiguration() { this = "EnvironmentToFileConfiguration" }
+    
+      override predicate isSource(DataFlow::Node source) {
+        exists(ExprNodes::ConstantReadAccessCfgNode env |
+          env.getExpr().getName() = "ENV" and
+          env = source.asExpr().(ExprNodes::ElementReferenceCfgNode).getReceiver()
+        )
+      }
+    
+      override predicate isSink(DataFlow::Node sink) {
+        sink = API::getTopLevelMember("File").getAMethodCall("open").getArgument(0)
+      }
+    }
+    
+    from EnvironmentToFileConfiguration config, DataFlow::Node environment, DataFlow::Node fileOpen
+    where config.hasFlow(environment, fileOpen)
+    select fileOpen, "This call to 'File.open' uses data from $@.", environment,
+      "an environment variable"
+
+Further reading
+---------------
+
+- ":ref:`Exploring data flow with path queries <exploring-data-flow-with-path-queries>`"
```

Comment:
Double-check is this link working?

Reply:
I've added a commit that I think will fix the PR check 🤞🏻